File: src/anchor/encrypted-container.ts
#! /usr/bin/env ts-node

/*
 * Description: Example of using the Keeta Anchor Client to create an encrypted container
 * Encrypted containers can only be decrypted by the specified recipient(s)
 */
import * as KeetaAnchor from '@keetanetwork/anchor';
import * as KeetaNet from '@keetanetwork/keetanet-client';

async function main() {
	// Create accounts
	const seed1 = KeetaNet.lib.Account.generateRandomSeed({ asString: true });
	const sender = KeetaNet.lib.Account.fromSeed(seed1, 0);

	const seed2 = KeetaNet.lib.Account.generateRandomSeed({ asString: true });
	const recipient = KeetaNet.lib.Account.fromSeed(seed2, 0);

	// Create sensitive data
	const privateData = {
		documentType: 'passport',
		documentNumber: 'X1234567',
		fullName: 'Alice Smith',
		dateOfBirth: '1990-01-01'
	};

	const plaintext = Buffer.from(JSON.stringify(privateData), 'utf-8');

	// Create an encrypted container that only the recipient can decrypt
	const container = KeetaAnchor.lib.EncryptedContainer.fromPlaintext(
		plaintext,
		[recipient],  // List of accounts that can decrypt
		{
			signer: sender,  // Optionally sign to prove authenticity
			locked: false    // Set to true to prevent adding more recipients later
		}
	);

	// Serialize the container for transmission
	const encodedContainer = await container.getEncodedBuffer();
	console.log('Encrypted container size:', encodedContainer.byteLength, 'bytes');

	// The recipient can decrypt the container
	const receivedContainer = KeetaAnchor.lib.EncryptedContainer.fromEncodedBuffer(
		encodedContainer,
		[recipient]  // Provide the recipient's account (with private key)
	);

	// Extract the plaintext
	const decryptedData = await receivedContainer.getPlaintext();
	const parsed = Buffer.from(decryptedData).toString('utf-8');

	console.log('Decrypted data:', parsed);

	// Verify the signature if it was signed
	if (receivedContainer.isSigned) {
		const signerPublicKey = receivedContainer.getSigningAccount()?.publicKeyString.get();
		console.log('Signed by:', signerPublicKey);
		console.log('Signature valid:', await receivedContainer.verifySignature());
	}
}

main().then(function() {
	process.exit(0);
}, function(err: unknown) {
	console.error(err);
	process.exit(1);
});

File: csharp/src/Anchor/EncryptedContainer.cs
using System.Text;
using System.Text.Json;
using KeetaNet.Anchor;
using KeetaNet.Anchor.Crypto;
using CryptoContainer = KeetaNet.Anchor.Crypto.EncryptedContainer;

namespace KeetaNet.Examples.Anchor;

public sealed class EncryptedContainerExample : IKeetaExample
{
	public string Id => "anchor/encrypted-container";

	public string Description =>
		"Example of using the Keeta Anchor Client to create an encrypted container";

	public Task<int> Run(string[] args, CancellationToken cancellationToken = default)
	{
		using var runtime = WasmRuntime.Load();

		// Create accounts.
		string seed1 = runtime.Accounts.GenerateRandomSeed();
		using Account sender = runtime.Accounts.FromSeed(seed1, 0, "ecdsa_secp256k1");

		string seed2 = runtime.Accounts.GenerateRandomSeed();
		using Account recipient = runtime.Accounts.FromSeed(seed2, 0, "ecdsa_secp256k1");

		// Create sensitive data.
		const string privateData = "sensitive information that only the recipient should see";
		byte[] plaintext = Encoding.UTF8.GetBytes(JsonSerializer.Serialize(privateData));

		// Create an encrypted container that only the recipient can decrypt.
		// The signer proves authenticity. Locked containers refuse new
		// recipients later.
		using CryptoContainer container = runtime.Containers.FromPlaintext(
			plaintext,
			principals: new[] { recipient },
			locked: false,
			signer: sender);

		// Serialize the container for transmission.
		byte[] encodedContainer = container.GetEncoded();
		Console.WriteLine($"Encrypted container size: {encodedContainer.Length} bytes");

		// The recipient can decrypt the container with its private key.
		using CryptoContainer receivedContainer = runtime.Containers.FromEncoded(
			encodedContainer,
			new[] { recipient });

		// Extract the plaintext.
		byte[] decryptedData = receivedContainer.GetPlaintext();
		string parsed = Encoding.UTF8.GetString(decryptedData);
		Console.WriteLine($"Decrypted data: {parsed}");

		// Verify the signature if it was signed.
		if (receivedContainer.IsSigned)
		{
			byte[]? signerKey = receivedContainer.GetSigningAccount();
			if (signerKey is not null)
			{
				using Account signerAccount = runtime.Accounts.FromPublicKeyAndType(
					Convert.ToHexString(signerKey));
				Console.WriteLine($"Signed by: {signerAccount.PublicKeyString}");
			}

			Console.WriteLine($"Signature valid: {receivedContainer.VerifySignature()}");
		}

		return Task.FromResult(0);
	}
}
